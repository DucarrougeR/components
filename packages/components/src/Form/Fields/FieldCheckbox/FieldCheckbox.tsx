import React, { forwardRef, Ref } from 'react'
import styled from 'styled-components'
import uuid from 'uuid/v4'
import { useFormContext } from '../../Form'
import { Checkbox, CheckboxProps } from '../../Inputs/Checkbox/Checkbox'
import { Field, FieldProps, omitFieldProps, pickFieldProps } from '../Field'

export interface FieldCheckboxProps extends FieldProps, CheckboxProps {}

const FieldCheckboxComponent = forwardRef(
  (props: FieldCheckboxProps, ref: Ref<HTMLInputElement>) => {
    const validationMessage = useFormContext(props)
    const { id = uuid() } = props
    return (
      <Field
        id={id}
        alignLabel="left"
        alignValidationMessage="right"
        validationMessage={validationMessage}
        {...pickFieldProps(props)}
      >
        <Checkbox
          {...omitFieldProps(props)}
          id={id}
          validationType={validationMessage && validationMessage.type}
          ref={ref}
        />
      </Field>
    )
  }
)

export const FieldCheckbox = styled(FieldCheckboxComponent)``
