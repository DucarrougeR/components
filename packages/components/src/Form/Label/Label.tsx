import styled from 'styled-components'
import {
  CompatibleHTMLProps,
  CustomizableAttributes,
  color,
  ColorProps,
  space,
  SpaceProps,
  textDecoration,
  TextDecorationProps,
  textTransform,
  TextTransformProps,
  typography,
  TypographyProps,
} from '@looker/design-tokens'

export const CustomizableLabelAttributes: CustomizableAttributes = {
  color: 'palette.charcoal800',
  fontSize: 'small',
  fontWeight: 'semiBold',
}

export interface LabelProps
  extends ColorProps,
    SpaceProps,
    TextDecorationProps,
    TextTransformProps,
    TypographyProps,
    CompatibleHTMLProps<HTMLLabelElement> {}

export const Label = styled.label<LabelProps>`
  color: ${() => CustomizableLabelAttributes.color};
  font-size: ${() => CustomizableLabelAttributes.fontSize};
  font-weight: ${() => CustomizableLabelAttributes.fontWeight};
  margin-right: ${props => props.theme.space.xsmall};
  ${color};
  ${space};
  ${textDecoration}
  ${textTransform};
  ${typography};
`
