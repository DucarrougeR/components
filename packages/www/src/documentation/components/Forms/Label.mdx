---
name: Label
route: /form/label
menu: Form & Inputs
---

import { ComponentDetail } from '../../components'

# Label

<ComponentDetail
  of="Label"
  status="experimental"
  figmaURL="file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
/>
