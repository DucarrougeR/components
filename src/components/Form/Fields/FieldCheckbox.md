
```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Fields/FieldCheckbox.tsx"
  feedbackTitle="FieldCheckbox Component Feedback" />
```


```js
  <FieldCheckbox id="theme-name" name="themeName" label="Theme Name" />
  <FieldCheckbox id="theme-name2" name="themeName2" label="Theme Name 2" />
  <FieldCheckbox id="theme-name3" name="themeName3" label="Theme Name 3" />
```
