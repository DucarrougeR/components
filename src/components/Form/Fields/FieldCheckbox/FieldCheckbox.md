
```js noeditor
const StatusAndResources = require('../../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Fields/FieldCheckbox.tsx"
  feedbackTitle="FieldCheckbox Component Feedback" />
```

### FieldCheckbox

The `<FieldCheckbox />` component has one required property: `label`. By default, the label will render to the left of the checkbox, which is the recommended convention, however this is adjustable with the `alignLabel` property.

Additionally, it is highly recommended to include the `name` property, as this will be used by forms to capture the inputted data and for validation.

The `required` property displays a red star by the label, indicating that the field must be filled out before submission.

```js
<FieldCheckbox name="box" label="Box" />
<FieldCheckbox name="box-2" label="Box 2" alignLabel="left" />
<FieldCheckbox name="requiredField" label="A required field" required />
```

### FieldCheckbox Validation

The `<FieldCheckbox />` components accepts `validationMessage` as a property. This is an object with properties `type` and `message`:
`
interface ValidationMessage {
  type: ValidationType
  message: string
}
`

`ValidationType` is a string literal, which accepts:
* error
* warning
* info
* success


If you have multiple fields in need of validation, it is recommended that you wrap them in a `<Form />` component and pass to it a dictionary with key as field name and value as a `ValidationMessage`.

The `alignValidationMessage` can also be used to adjust where the validation message appears in relation to the input field. By default, it will render to the right, which is the best practice in most cases.

___Note: In the current implementation, any `ValidationMessage` set in a `<FieldCheckbox />` component will be overridden by `ValidationMessages` set in the parent `<Form />` component. Because of this, it is highly recommended that you only ever directly set `ValidationMessage` if a `<FieldCheckbox />` is outside a `<Form />`.___

```js
<FieldCheckbox name="someBox" label="Some Box" validationMessage={{type: "error", message: "This is an error"}}/>
<FieldCheckbox name="some-box-2" label="Some Box 2" validationMessage={{type: "warning", message: "Here is a warning"}} alignValidationMessage="bottom" />
```
