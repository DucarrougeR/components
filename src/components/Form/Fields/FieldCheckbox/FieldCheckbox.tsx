import * as React from 'react'
import uuid from 'uuid/v4'
import { styled } from '../../../../style'
import { withForm } from '../../Form'
import { Checkbox, CheckboxProps } from '../../Inputs/Checkbox/Checkbox'
import { Field, FieldProps } from '../Field'

interface FieldCheckboxProps extends FieldProps, CheckboxProps {}

const InternalFieldCheckbox = (props: FieldCheckboxProps) => {
  const {
    alignLabel,
    alignValidationMessage,
    label,
    labelWidth,
    labelFontWeight,
    validationMessage,
    id = uuid(),
    ...checkboxProps
  } = props
  return (
    <Field
      {...props}
      id={id}
      alignLabel={alignLabel ? alignLabel : 'left'}
      alignValidationMessage={
        alignValidationMessage ? alignValidationMessage : 'right'
      }
    >
      <Checkbox
        id={id}
        {...checkboxProps}
        validationType={validationMessage && validationMessage.type}
      />
    </Field>
  )
}

const FieldCheckboxFactory = React.forwardRef((props: FieldProps, ref) => (
  <InternalFieldCheckbox
    innerRef={ref as React.RefObject<HTMLElement>}
    {...props}
  />
))

export const FieldCheckbox = styled(withForm(FieldCheckboxFactory))``
