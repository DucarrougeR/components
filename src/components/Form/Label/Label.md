```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/relens/blob/master/src/components/Form/Label/Label.tsx"
  feedbackTitle="Label Component Feedback" />
```
