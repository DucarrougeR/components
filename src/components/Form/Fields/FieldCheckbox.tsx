import * as React from 'react'
import { FormControl, FormControlDirections } from '../FormGroup/FormGroup'
import { InputCheckbox } from '../Inputs/InputCheckbox'
import { Label } from '../Label/Label'

export interface FieldCheckboxProps {
  id: string
  label: string
  name: string
}

export const FieldCheckbox: React.SFC<FieldCheckboxProps> = ({
  id,
  label,
  name,
  ...props
}) => {
  return (
    <FormControl alignLabel={FormControlDirections.Left} alignCenter>
      <Label htmlFor={id}>{label}</Label>
      <InputCheckbox id={id} name={name} {...props} />
    </FormControl>
  )
}
