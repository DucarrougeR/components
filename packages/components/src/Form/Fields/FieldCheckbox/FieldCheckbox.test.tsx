import 'jest-styled-components'
import * as React from 'react'
import { createWithTheme, mountWithTheme } from '@looker/components-test-utils'
import { Label } from '../../Label/Label'
import { FieldCheckbox } from './FieldCheckbox'

test('A FieldCheckbox', () => {
  const component = createWithTheme(
    <FieldCheckbox label="👍" name="thumbsUp" id="thumbs-up" />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A FieldCheckbox with checked value', () => {
  const component = createWithTheme(
    <FieldCheckbox label="👍" name="thumbsUp" id="thumbs-up" checked />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A required FieldCheckbox', () => {
  const component = createWithTheme(
    <FieldCheckbox label="👍" name="thumbsUp" id="thumbs-up" required />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A FieldCheckbox with an error validation aligned to the bottom', () => {
  const component = mountWithTheme(
    <FieldCheckbox
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      validationMessage={{ type: 'error', message: 'This is an error' }}
      alignValidationMessage="bottom"
    />
  )
  expect(component.find(Label).props().htmlFor).toEqual(
    component.find('input').props().id
  )
})
