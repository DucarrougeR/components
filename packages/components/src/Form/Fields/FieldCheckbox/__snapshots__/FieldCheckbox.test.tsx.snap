// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FieldCheckbox 1`] = `
.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-right: 0.5rem;
  color: #262D33;
  font-weight: 600;
  font-size: 0.875rem;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c1 .c2 {
  text-align: right;
  width: 20%;
}

<div
  className="c0 c1 "
>
  <label
    className="c2 c3"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c0"
  >
    <div
      className="c4"
    >
      <input
        className="c5 "
        id="thumbs-up"
        name="thumbsUp"
        type="checkbox"
      />
    </div>
  </div>
</div>
`;

exports[`A FieldCheckbox with checked value 1`] = `
.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-right: 0.5rem;
  color: #262D33;
  font-weight: 600;
  font-size: 0.875rem;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c1 .c2 {
  text-align: right;
  width: 20%;
}

<div
  className="c0 c1 "
>
  <label
    className="c2 c3"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c0"
  >
    <div
      className="c4"
    >
      <input
        checked={true}
        className="c5 "
        id="thumbs-up"
        name="thumbsUp"
        type="checkbox"
      />
    </div>
  </div>
</div>
`;

exports[`A required FieldCheckbox 1`] = `
.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-right: 0.5rem;
  color: #262D33;
  font-weight: 600;
  font-size: 0.875rem;
}

.c6 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c4 {
  color: #990F14;
}

.c1 .c2 {
  text-align: right;
  width: 20%;
}

<div
  className="c0 c1 "
>
  <label
    className="c2 c3"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
    htmlFor="thumbs-up"
  >
    üëç
    <span
      aria-hidden="true"
      className="c4"
    >
       
      *
    </span>
  </label>
  <div
    className="c0"
  >
    <div
      className="c5"
    >
      <input
        className="c6 "
        id="thumbs-up"
        name="thumbsUp"
        required={true}
        type="checkbox"
      />
    </div>
  </div>
</div>
`;
